### PR TITLE
Use invalidation rectangle

### PR DESCRIPTION
**Pull request by raphlinus**
Do partial invalidation of EditView rather than redrawing the whole
screen. Should improve latency and power usage.

Also changes blinking cursor to not draw when off, rather than draw
background color.

### CODE REVIEW

File: XiEditor/EditView.swift [hunk 1]
```diff
@@ -436,7 +444,7 @@ class EditView: NSView, NSTextInputClient {
     /// timer callback to toggle the blink state
     @objc func _blinkInsertionPoint() {
         _cursorStateOn = !_cursorStateOn
-        needsDisplay = true
+        partialInvalidate(invalid: dataSource.lines.cursorInval)
```
Comment (cmyr) [inline XiEditor/EditView.swift]: A minor improvement possible here: `EditView` could just have a `invalidationSetForCurrentCursor` property that is set when the update arrives, which would save recomputing on each blink.

Not a blocker, more an observation.

File: XiEditor/LineCache.swift [hunk 2]
```diff
@@ -178,4 +199,32 @@ class LineCache {
...
+                inval.addRange(start: i + nInvalidBefore, n: 1)
+            }
+        }
+        return inval
+    }
+}
+
+/// A set of line numbers to be invalidated, in run-length representation
+class InvalSet {
+    var ranges: [(Int, Int)] = []
```
Comment (cmyr) [inline XiEditor/LineCache.swift]: Non-block: I'd be inclined to use either the `Range` or `ClosedRange` foundation type here.

File: XiEditor/EditView.swift (hunk 1 above)
Comment (raphlinus) [reply XiEditor/EditView.swift]: I agree this is a good optimization, but I expect the impact to be tiny.

**Comment by raphlinus** [comment]
I also changed it so InvalSet is ranges. Agree this is nicer, a pair always has ambiguity whether it's (start, count) or (start, end).

File: XiEditor/EditView.swift (hunk 1 above)
Comment (cmyr) [reply XiEditor/EditView.swift]: agreed, just an observation. :)

**Pull request merged**